src/posix/platform/infra_if.cpp

```diff
@@ -250,21 +289,97 @@ void platformInfraIfDeinit(void)
         sInfraIfIcmp6Socket = -1;
     }
 
+    if (sNetLinkSocket != -1)
+    {
+        close(sNetLinkSocket);
+        sNetLinkSocket = -1;
+    }
+
     sInfraIfIndex = 0;
 }
 
 void platformInfraIfUpdateFdSet(fd_set &aReadFdSet, int &aMaxFd)
 {
     VerifyOrExit(sInfraIfIcmp6Socket != -1);
+    VerifyOrExit(sNetLinkSocket != -1);
 
     FD_SET(sInfraIfIcmp6Socket, &aReadFdSet);
     aMaxFd = OT_MAX(aMaxFd, sInfraIfIcmp6Socket);
 
+    FD_SET(sNetLinkSocket, &aReadFdSet);
+    aMaxFd = OT_MAX(aMaxFd, sNetLinkSocket);
+
 exit:
     return;
 }
 
-void platformInfraIfProcess(otInstance *aInstance, const fd_set &aReadFdSet)
+// Create a net-link socket that subscribes to link & addresses events.
+static int CreateNetLinkSocket(void)
+{
+    int                sock;
+    int                rval;
+    struct sockaddr_nl addr;
+
+    sock = socket(AF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
+    VerifyOrDie(sock != -1, OT_EXIT_ERROR_ERRNO);
+
+    memset(&addr, 0, sizeof(addr));
+    addr.nl_family = AF_NETLINK;
+    addr.nl_groups = RTMGRP_LINK | RTMGRP_IPV6_IFADDR;
+    addr.nl_pid    = getpid();
+
+    rval = bind(sock, reinterpret_cast<struct sockaddr *>(&addr), sizeof(addr));
+    VerifyOrDie(rval == 0, OT_EXIT_ERROR_ERRNO);
+
+    return sock;
+}
+
+static void ReceiveNetLinkMessage(otInstance *aInstance)
+{
+    const size_t kMaxNetLinkBufSize = 8192;
+    ssize_t      len;
+    union
+    {
+        nlmsghdr mHeader;
+        uint8_t  mBuffer[kMaxNetLinkBufSize];
```

Comment:
This buffer is huge. Do we really need such large buffer?

Reply:
Yes.